### DISCUSSION TITLE
solving PDE for convection diffusion

### DISCUSSION THREAD

**Discussion by ymharshe** [zwicker-group/py-pde#606] [Q&A]
I am solving classical convection-diffusion problem which takes the following form.

![image](https://github.com/user-attachments/assets/9ead24a6-2f66-46af-9c6c-0abb1c14679b)

with B.C. 

![image](https://github.com/user-attachments/assets/c5ea2b9d-f04f-4d3f-adff-c6ff5382bd5a)

for the moment I am taking constant D and U

My code looks like as follows.

```python
from pde import PDE, CartesianGrid, MemoryStorage, ScalarField, plot_kymograph

from pde import PDE, CartesianGrid, MemoryStorage, ScalarField, plot_kymograph

diffusivity = 1e-9
velocity = 1e-6
cmax = 0.6

term_1 = f"({diffusivity}) * laplace(c)"
term_2 = f"({velocity}) * gradient(c)"

bc_lower = {"type": "mixed", "value": -velocity/diffusivity, "const": 0}
bc_upper = {"type": "mixed", "value": -velocity/diffusivity, "const": 0}

eq = PDE({"c": f"{term_1} - {term_2}"}, bc=[bc_lower, bc_upper])

grid = CartesianGrid([[0, 0.08]], 8000)  # generate grid
field = ScalarField(grid, 0.2)  # generate initial condition


storage = MemoryStorage()  # store intermediate information of the simulation
res = eq.solve(field, 86400, dt=1e-1, tracker=storage.tracker(1))  # solve the PDE

plot_kymograph(storage)  # visualize the result in a space-time plot
```

Can anyone tell me if this implementation is correct or not? Also, I get some errors while running the code.

![image](https://github.com/user-attachments/assets/00c2af68-f330-4bdd-bd7a-edb47d466dca)

![image](https://github.com/user-attachments/assets/3c72f91d-75bf-44ed-a052-974181dae582)

**Comment by ymharshe** [comment]
The code above, does not have the
![image](https://github.com/user-attachments/assets/a873dc53-95bb-49ce-9bf0-d19ce0ca6184)

term. 
As I wanted to verify that the code works without this term. But then I would like to eventually have the code including this term as mentioned in the first post here.

Thanks in advance for your feedback and support.
Regards,
Yogesh

**Reply by david-zwicker** [reply]
As far as I can see, the problem is that your code is dimensionally inconsistent since you're basically adding the laplacian of the field (a scalar) to the gradient of the field (a vector). A simple fix could be to replace `gradient` by `d_dx` if you want to stick with 1D problems. I unfortunately can't test this right now, but I suspect that this is the fixes the problem described above.

**Reply by ymharshe** [reply]
Thanks for the feedback. I think this the reason and helps me.  

term_1 = f"({diffusivity}) * d2_dx2(c)"
term_2 = f"({velocity}) * d_dx(c) * d_dx(1-c/{cmax})"
eq = PDE({"c": f"{term_1} - {term_2}"}, bc=[bc_lower, bc_upper])

However, I am still struggling to implement the boundary condition.  The actual boundary condition is 

![image](https://github.com/user-attachments/assets/721255cb-7397-448b-bf62-39c9e0efb322)



and I tried to implement it via following code but it does not work.

bc_lower = {"type": "mixed_expression", "value_expression": -velocity/diffusivity*(1-sum(c)/{cmax}), "const": 0}
bc_upper = {"type": "mixed_expression", "value_expression": -velocity/diffusivity*(1-sum(c)/{cmax}), "const": 0}

but it does not work!

How can I implement the boundary condition of the following type?

![image](https://github.com/user-attachments/assets/72bc006c-7430-46b0-96b0-2bdf47a4c2e3)

**Reply by ymharshe** [reply]
Any inputs or help from anyone in setting up Bounday conditions for the PDE?

**Reply by david-zwicker** [reply]
I don't understand the condition fully. Is the sum supposed to be an integral in the boundary condition? If so, this is not currently supported by py-pde. You could open an feature request on this, but I currently don't have the bandwidth to work on this.

**Reply by ymharshe** [reply]
The boundary conditions is such that  there is a term in the bracket which has a summation of all dependent variable (phi at each z). phi is   matrix n entities in z spaces. Please see,  if it can be handled by py-pde.

**Reply by david-zwicker** [reply]
It sounds as if you want to have an integral over the field in the boundary condition, which is currently not supported.

**Reply by ymharshe** [reply]
it is like system of PDEs solved simultaneously.  Can py pde handle a set of PDEs solved simultaneously with different initial conditions? The boundary condition says that the space derivative of dependent variables of the system of PDEs at both boundaries follow this condition. The sum is the sum of dependent variables at the boundaries.

**Reply by david-zwicker** [reply]
py-pde supports systems of PDEs, but coupling these equations via boundary conditions is not supported out of the box. You could develop a work-around by using [the advanced `UserBC` class](https://py-pde.readthedocs.io/en/latest/packages/pde.grids.boundaries.local.html#pde.grids.boundaries.local.UserBC), which allows you to set virtual points yourself. This requires quite detailed knowledge of the internal workings and I unfortunately don't have the bandwidth to help you with the implementation.

**Reply by ymharshe** [reply]
is there any example of solving system, of PDEs with individual initial conditions and boundary conditions? Actually, the BC I have is similar to the Robin condition where the space derivative is equal to some constant multiplied by the dependent variable  and (1- sum of all dependent variable values at the boundary conditions/ constant)
